src/device/mmio.rs

```diff
@@ -0,0 +1,355 @@
+// a `VirtioDevice`, provided the status check is successful.
+// TODO: This function and its uses will likely have to be updated when we start offering
+// packed virtqueue support as well.
+fn update_queue_field<M, D, F>(device: &mut D, f: F)
+where
+    M: GuestAddressSpace,
+    D: WithDriverSelect<M> + ?Sized,
+    F: FnOnce(&mut Queue<M>),
+{
+    if device.check_device_status(status::FEATURES_OK, status::DRIVER_OK | status::FAILED) {
+        if let Some(queue) = device.selected_queue_mut() {
+            f(queue);
+        } else {
+            warn!("update invalid virtio queue");
+        }
+    } else {
+        warn!(
+            "update virtio queue in invalid state 0x{:x}",
+            device.device_status()
+        );
+    }
+}
+
+// Helper function that rewrites the most significant 4 bytes of the provided `GuestAddress`.
+fn set_high(v: &mut GuestAddress, hi: u32) {
+    *v = (*v & 0xffff_ffff) | (u64::from(hi) << 32)
+}
+
+// Helper function that rewrites the least significant 4 bytes of the provided `GuestAddress`.
+fn set_low(v: &mut GuestAddress, lo: u32) {
+    *v = (*v & !0xffff_ffff) | u64::from(lo)
+}
+
+/// A common interface for Virtio devices that use the MMIO transport, which also provides a
+/// default implementation of read and write operations from/to the device registers and
+/// configuration space.
+// Adding the `M` generic parameter that's also required by `VirtioDevice` for the time being.
+pub trait VirtioMmioDevice<M: GuestAddressSpace>: WithDriverSelect<M> {
+    /// Callback invoked when the driver writes a value to the Queue Notify configuration register.
+    ///
+    /// This is the simplest mechanism the driver can use to notify a virtio MMIO device. The
+    /// meaning of the value is interpreted as specified by the standard. Many VMMs use something
+    /// like the KVM `ioeventfd` mechanism, which eliminates the need to implement this method.
+    // TODO: Move this to the `VirtioDevice` interface if it uniformly applies to all transports.
+    fn queue_notify(&mut self, _val: u32) {
+        // Do nothing by default.
+    }
+
+    /// Handle a driver read operation from the MMIO space of the device.
+    fn read(&self, offset: u64, data: &mut [u8]) {
+        match offset {
+            // The standard specifies that accesses to configuration registers are 32-bit wide.
+            0x00..=0xff if data.len() == 4 => {
+                let v = match offset {
+                    0x0 => MMIO_MAGIC_VALUE,
+                    0x04 => MMIO_VERSION,
+                    0x08 => self.device_type(),
+                    0x0c => VENDOR_ID,
+                    0x10 => match self.device_features_select() {
+                        0 => self.device_features() as u32,
+                        1 => (self.device_features() >> 32) as u32,
+                        // No device features defined beyond the first two pages.
+                        _ => 0,
```

Comment:
This can be tested too fairly easily if the `Dummy` device does `d.set_device_features_select(2)` before an MMIO read from offset `0x10`